src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,275 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2024 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/apply.h>
+#include <xrpld/app/tx/detail/Batch.h>
+#include <xrpld/ledger/Sandbox.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/basics/Log.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/TER.h>
+#include <xrpl/protocol/TxFlags.h>
+
+namespace ripple {
+
+TxConsequences
+Batch::makeTxConsequences(PreflightContext const& ctx)
+{
+    return TxConsequences{ctx.tx, TxConsequences::normal};
+}
+
+NotTEC
+Batch::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureBatch))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+
+    if (tx.getFlags() & tfBatchMask)
+    {
+        JLOG(ctx.j.warn()) << "Batch: invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    AccountID const outerAccount = tx.getAccountID(sfAccount);
+
+    auto const& txns = tx.getFieldArray(sfRawTransactions);
+    STVector256 const hashes = tx.getFieldV256(sfTxIDs);
```

Comment:
is there a check that verifies the each hashes map exactly to each raw transaction? didn't find it here, but perhaps its somewhere else that i may have not cross my way yet

Reply:
Yes in preflight. https://github.com/XRPLF/rippled/pull/5060/files/1367e03543171c47af68bc93e370f99e256bcb41#diff-a57b3ccde9d91552e7a3d76d3ab2aa35fbd130ceac001582b675c065a9ab643dR80